File: leetcode_go/problems/0002_add_two_numbers/add_two_numbers.go
package problem0002

import l "github.com/swissarmybox/leetcode_go/helpers/linked_list"

type ListNode = l.ListNode

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := &ListNode{Val: -1}
	node := dummy

	extra := 0

	for l1 != nil && l2 != nil {
		sum := l1.Val + l2.Val + extra
		l1 = l1.Next
		l2 = l2.Next
		extra = 0

		if sum >= 10 {
			sum = sum - 10
			extra = 1
		}

		node.Next = &ListNode{Val: sum}
		node = node.Next
	}

	for l1 != nil {
		sum := l1.Val + extra
		l1 = l1.Next
		extra = 0

		if sum >= 10 {
			sum = sum - 10
			extra = 1
		}

		node.Next = &ListNode{Val: sum}
		node = node.Next
	}

	for l2 != nil {
		sum := l2.Val + extra
		l2 = l2.Next
		extra = 0

		if sum >= 10 {
			sum = sum - 10
			extra = 1
		}

		node.Next = &ListNode{Val: sum}
		node = node.Next
	}

	if extra > 0 {
		node.Next = &ListNode{Val: extra}
	}

	return dummy.Next
}

File: leetcode_js/src/addtwoNumbers.js
const { ListNode } = require('../tests/helpers/listNode');

const addTwoNumbers = function (l1, l2) {
  const head = new ListNode(null);
  let node = head;
  let additional = 0;

  while (l1 !== null || l2 !== null) {
    if (l1 !== null && l2 !== null) {
      const sum = l1.val + l2.val + additional;

      if (sum >= 10) {
        additional = 1;
        node.next = new ListNode(sum - 10);
      } else {
        additional = 0;
        node.next = new ListNode(sum);
      }

      l1 = l1.next;
      l2 = l2.next;
    } else if (l1 !== null) {
      const sum = l1.val + additional;
      if (sum >= 10) {
        additional = 1;
        node.next = new ListNode(sum - 10);
      } else {
        additional = 0;
        node.next = new ListNode(sum);
      }

      l1 = l1.next;
    } else {
      const sum = l2.val + additional;
      if (sum >= 10) {
        additional = 1;
        node.next = new ListNode(sum - 10);
      } else {
        additional = 0;
        node.next = new ListNode(sum);
      }

      l2 = l2.next;
    }

    node = node.next;
  }

  if (additional === 1) {
    node.next = new ListNode(1);
  }

  return head.next;
};

module.exports = { addTwoNumbers };
